File: python/src/linodemcp/linode/metrics.py
"""Linode API-request metric recording wired through a context variable.

The server injects a recorder into the dispatch context so the client records
each API round trip without importing observability directly. Mirrors the Go
linode/metrics.go context wiring.
"""

import contextvars
from typing import Protocol


class APIRecorder(Protocol):
    """Records metrics for one Linode API HTTP round trip."""

    def record_api_request(
        self, endpoint: str, method: str, status: int, duration_seconds: float
    ) -> None:
        """Record a single API request."""
        ...


_api_recorder: contextvars.ContextVar[APIRecorder | None] = contextvars.ContextVar(
    "linode_api_recorder", default=None
)


def set_api_recorder(
    recorder: APIRecorder | None,
) -> contextvars.Token[APIRecorder | None]:
    """Bind the recorder for the current context; returns a reset token."""
    return _api_recorder.set(recorder)


def reset_api_recorder(token: contextvars.Token[APIRecorder | None]) -> None:
    """Restore the recorder bound before the matching set_api_recorder."""
    _api_recorder.reset(token)


def get_api_recorder() -> APIRecorder | None:
    """Return the recorder bound for the current context, or None."""
    return _api_recorder.get()


def metrics_endpoint(endpoint: str) -> str:
    """Strip the query string so the endpoint metric label stays path-level.

    Path IDs still vary, so the endpoint label is higher cardinality than
    method or status; tighten to route templates if series count becomes a
    concern.
    """
    return endpoint.split("?", 1)[0]

File: go/internal/linode/metrics.go
package linode

import (
	"context"
	"strings"
)

// APIRecorder records metrics for a single Linode API HTTP round trip. The
// server injects an implementation into the request context (see
// WithAPIRecorder) so the client records without taking a direct dependency
// on the observability package. *observability.Observability satisfies it
// through its RecordAPIRequest method.
type APIRecorder interface {
	RecordAPIRequest(ctx context.Context, endpoint, method string, status int, duration float64)
}

type apiRecorderKey struct{}

// WithAPIRecorder returns a context carrying the recorder the client uses to
// report each API request. A nil recorder is ignored so callers can pass a
// not-yet-wired recorder without branching.
func WithAPIRecorder(ctx context.Context, recorder APIRecorder) context.Context {
	if recorder == nil {
		return ctx
	}

	return context.WithValue(ctx, apiRecorderKey{}, recorder)
}

// apiRecorderFromContext returns the recorder set by WithAPIRecorder, or nil
// when none is present (the default, so a client used outside the server
// dispatch path records nothing rather than crashing).
func apiRecorderFromContext(ctx context.Context) APIRecorder {
	recorder, _ := ctx.Value(apiRecorderKey{}).(APIRecorder)

	return recorder
}

// metricsEndpoint trims the query string so the endpoint metric label stays
// at the path level. Path IDs still vary, so the endpoint label is higher
// cardinality than method or status; tighten it to route templates if the
// series count becomes a memory concern.
func metricsEndpoint(endpoint string) string {
	if before, _, ok := strings.Cut(endpoint, "?"); ok {
		return before
	}

	return endpoint
}
